plutus-core/testlib/PlutusCore/Generators/PIR/GenTm.hs

```diff
@@ -0,0 +1,317 @@
+-- editorconfig-checker-disable-file
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE UndecidableInstances  #-}
+
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module PlutusCore.Generators.PIR.GenTm
+  ( module PlutusCore.Generators.PIR.GenTm
+  , module Export
+  , Arbitrary (..)
+  , Gen
+  ) where
+
+import Control.Monad.Except
+import Control.Monad.Reader
+
+import Data.List.NonEmpty (NonEmpty (..))
+import Data.Map.Strict (Map)
+import Data.Map.Strict qualified as Map
+import Data.Maybe
+import Data.Set (Set)
+import Data.Set qualified as Set
+import Data.Set.Lens (setOf)
+import Data.String
+import Test.QuickCheck (Arbitrary (..), Gen)
+import Test.QuickCheck qualified as QC
+import Test.QuickCheck.GenT as Export hiding (var)
+
+import PlutusCore.Default
+import PlutusCore.Name
+import PlutusIR
+import PlutusIR.Compiler
+import PlutusIR.Subst
+
+import PlutusCore.Generators.PIR.Common
+import PlutusCore.Generators.PIR.Utils
+
+instance MonadReader r m => MonadReader r (GenT m) where
+    ask = lift ask
+    local f (GenT k) = GenT $ \qc size -> local f $ k qc size
+
+-- | Term generators carry around a context to know
+-- e.g. what types and terms are in scope.
+type GenTm = GenT (Reader GenEnv)
+
+data GenEnv = GenEnv
+  { geSize               :: Int
+  -- ^ Generator size bound. See Note [Recursion Control and geSize]
+  , geDatas              :: Map TyName (Datatype TyName Name DefaultUni ())
```

Comment:
why does only `TypeCtx` get an alias?

Reply:
There's `TypeSub` too. Dunno, I didn't find aliasing this one too important, but then I didn't review the terms part, which is where this one is used. I'm going to postpone it until I start dealing with terms.